TrigScint/src/TrigScint/Firmware/hitproducer_hw.cxx

```diff
@@ -0,0 +1,144 @@
+#pragma HLS INTERFACE ap_fifo depth = 16 port = FIFO[47]
+
+#pragma HLS INTERFACE ap_fifo depth = 16 port = FIFO[48]
+#pragma HLS INTERFACE ap_fifo depth = 16 port = FIFO[49]
+
+#pragma HLS PIPELINE
+
+  /// Indices of first bin of each subrange
+  ap_uint<14> nbins_[5] = {0, 16, 36, 57, 64};
+
+  /// Charge lower limit of all the 16 subranges
+  ap_uint<14> edges_[17] = {0,     34,    158,    419,    517,   915,
+                            1910,  3990,  4780,   7960,   15900, 32600,
+                            38900, 64300, 128000, 261000, 350000};
+  /// sensitivity of the subranges (Total charge/no. of bins)
+  ap_uint<14> sense_[16] = {3,   6,   12,  25,   25,   50,   99,   198,
+                            198, 397, 794, 1587, 1587, 3174, 6349, 12700};
+
+  for (int i = 0; i < NHITS; i++) {
+    outHit[i].bID = -1;
+    outHit[i].mID = 0;
+    outHit[i].Time = 0;
+    outHit[i].Amp = 0;
+    ap_uint<14> word1 = FIFO[i][0];
+    ap_uint<14> word2 = FIFO[i][1];
+    ap_uint<14> word3 = FIFO[i][2];
+    ap_uint<14> word4 = FIFO[i][3];
+    ap_uint<14> word5 = FIFO[i][4];
+    ap_uint<14> word6 = FIFO[i][5];
+    ap_uint<16> charge1;
+    ap_uint<16> charge2;
+    ap_uint<16> charge3;
+    ap_uint<16> charge4;
+    ap_uint<16> charge5;
+    ap_uint<16> charge6;
+
+    ap_uint<14> rr = (word1 >> 6) / 64;
+    ap_uint<14> v1 = (word1 >> 6) % 64;
+    ap_uint<14> ss =
+        1 * (v1 > nbins_[1]) + 1 * (v1 > nbins_[2]) + 1 * (v1 > nbins_[3]);
+    charge1 = edges_[4 * rr + ss] + (v1 - nbins_[ss]) * sense_[4 * rr + ss] +
+              sense_[4 * rr + ss] / 2 - 1;
+
+    rr = (word2 >> 6) / 64;
+    v1 = (word2 >> 6) % 64;
+    ss = 1 * (v1 > nbins_[1]) + 1 * (v1 > nbins_[2]) + 1 * (v1 > nbins_[3]);
+    charge2 = edges_[4 * rr + ss] + (v1 - nbins_[ss]) * sense_[4 * rr + ss] +
+              sense_[4 * rr + ss] / 2 - 1;
+
+    rr = (word3 >> 6) / 64;
+    v1 = (word3 >> 6) % 64;
+    ss = 1 * (v1 > nbins_[1]) + 1 * (v1 > nbins_[2]) + 1 * (v1 > nbins_[3]);
+    charge3 = edges_[4 * rr + ss] + (v1 - nbins_[ss]) * sense_[4 * rr + ss] +
+              sense_[4 * rr + ss] / 2 - 1;
+
+    rr = (word4 >> 6) / 64;
+    v1 = (word4 >> 6) % 64;
+    ss = 1 * (v1 > nbins_[1]) + 1 * (v1 > nbins_[2]) + 1 * (v1 > nbins_[3]);
+    charge4 = edges_[4 * rr + ss] + (v1 - nbins_[ss]) * sense_[4 * rr + ss] +
+              sense_[4 * rr + ss] / 2 - 1;
+
+    rr = (word5 >> 6) / 64;
+    v1 = (word5 >> 6) % 64;
+    ss = 1 * (v1 > nbins_[1]) + 1 * (v1 > nbins_[2]) + 1 * (v1 > nbins_[3]);
+    charge5 = edges_[4 * rr + ss] + (v1 - nbins_[ss]) * sense_[4 * rr + ss] +
+              sense_[4 * rr + ss] / 2 - 1;
+
+    rr = (word6 >> 6) / 64;
+    v1 = (word6 >> 6) % 64;
+    ss = 1 * (v1 > nbins_[1]) + 1 * (v1 > nbins_[2]) + 1 * (v1 > nbins_[3]);
+    charge6 = edges_[4 * rr + ss] + (v1 - nbins_[ss]) * sense_[4 * rr + ss] +
+              sense_[4 * rr + ss] / 2 - 1;
+
+    outHit[i].bID = i;
+    outHit[i].Time = (word1 & 63);
```

Comment:
is this 63 on purpose? everything above was with 64. Maybe add a comment in the code where these numbers are coming from

Reply:
I put a comment; it is needed to extract the tdc alone.